File: src/make.ts
// `bplan make`: baut das Modell und schreibt eine eigenständige HTML-Datei.
//
// Das Frontend kommt als Text-Import herein und gehört damit zum Modulgraphen:
// läuft `bplan` über eine URL, lädt Deno die HTML-Datei mit und legt sie in
// denselben Cache wie den Code. `make` ersetzt darin den Platzhalter durch das
// Modell als JSON. Die entstehende Datei ist per Mail verschickbar und
// funktioniert per Doppelklick ohne Server.

import FRONTEND from "../frontend/index.html" with { type: "text" };

import type { Modell } from "./model.ts";

/** Stelle im Frontend, an die das Modell kommt. */
const PLATZHALTER = "<!--MODEL-->";

/** Setzt das Modell in das Frontend ein. */
export function seite(modell: Modell): string {
  if (!FRONTEND.includes(PLATZHALTER)) {
    throw new Error(`im Frontend fehlt der Platzhalter ${PLATZHALTER}`);
  }
  return FRONTEND.replaceAll(PLATZHALTER, fuerScriptTag(JSON.stringify(modell)));
}

/** Das Frontend, wie `serve` es unverändert ausliefert. */
export function frontend(): string {
  return FRONTEND;
}

/**
 * Ersetzt jedes `<` durch seine JSON-Escape-Form. In der Ausgabe von
 * `JSON.stringify` steht `<` nur innerhalb von Zeichenketten, das Ergebnis
 * bleibt also gültiges JSON. Damit kann keine Anmerkung mit einem
 * schließenden Script-Tag das eingebettete Modell vorzeitig beenden.
 */
function fuerScriptTag(json: string): string {
  return json.replaceAll("<", "\\u003c");
}

export function groesse(bytes: number): string {
  if (bytes < 1024) return `${bytes} Bytes`;
  if (bytes < 1024 * 1024) return `${Math.floor(bytes / 1024)} KB`;
  return `${(bytes / (1024 * 1024)).toFixed(1)} MB`;
}

/** Legt fehlende Zwischenverzeichnisse an und schreibt die Datei. */
export function schreibe(ausgabe: string, html: string): number {
  const ordner = ausgabe.replace(/\/+[^/]*$/, "");
  if (ordner !== "" && ordner !== ausgabe) {
    Deno.mkdirSync(ordner, { recursive: true });
  }
  const bytes = new TextEncoder().encode(html);
  Deno.writeFileSync(ausgabe, bytes);
  return bytes.length;
}

File: src/make.rs
//! `bplan make`: baut das Modell und schreibt eine eigenständige HTML-Datei.
//!
//! Das Frontend ist zur Bauzeit eingebettet. `make` ersetzt darin den
//! Platzhalter durch das Modell als JSON. Die entstehende Datei ist per Mail
//! verschickbar und funktioniert per Doppelklick ohne Server.

use std::path::Path;
use std::process::ExitCode;

use crate::model::Modell;
use crate::validate;

/// Das komplette Frontend, siehe SPEC.md, Abschnitt "Frontend".
const FRONTEND: &str = include_str!("../frontend/index.html");

/// Stelle im Frontend, an die das Modell kommt.
const PLATZHALTER: &str = "<!--MODEL-->";

pub fn baue(pfad: &Path, ausgabe: &Path) -> ExitCode {
    let ergebnis = validate::pruefe(pfad);
    crate::zeige(&ergebnis);
    if ergebnis.hat_fehler() {
        eprintln!("bplan make: abgebrochen, erst die Fehler beheben.");
        return ExitCode::FAILURE;
    }

    let html = match seite(&ergebnis.modell) {
        Ok(html) => html,
        Err(meldung) => {
            eprintln!("bplan make: {meldung}");
            return ExitCode::FAILURE;
        }
    };

    if let Some(ordner) = ausgabe.parent() {
        if !ordner.as_os_str().is_empty() {
            if let Err(fehler) = std::fs::create_dir_all(ordner) {
                eprintln!(
                    "bplan make: Verzeichnis {} laesst sich nicht anlegen: {fehler}",
                    ordner.display()
                );
                return ExitCode::FAILURE;
            }
        }
    }

    if let Err(fehler) = std::fs::write(ausgabe, &html) {
        eprintln!(
            "bplan make: {} laesst sich nicht schreiben: {fehler}",
            ausgabe.display()
        );
        return ExitCode::FAILURE;
    }

    println!("{} geschrieben, {}", ausgabe.display(), groesse(html.len()));
    ExitCode::SUCCESS
}

/// Setzt das Modell in das eingebettete Frontend ein.
fn seite(modell: &Modell) -> Result<String, String> {
    if !FRONTEND.contains(PLATZHALTER) {
        return Err(format!("im Frontend fehlt der Platzhalter {PLATZHALTER}"));
    }
    let json = serde_json::to_string(modell)
        .map_err(|fehler| format!("Modell laesst sich nicht serialisieren: {fehler}"))?;
    Ok(FRONTEND.replace(PLATZHALTER, &fuer_script_tag(&json)))
}

/// Ersetzt jedes `<` durch seine JSON-Escape-Form. In der Ausgabe von
/// `serde_json` steht `<` nur innerhalb von Zeichenketten, das Ergebnis bleibt
/// also gültiges JSON. Damit kann keine Anmerkung mit einem schließenden
/// Script-Tag das eingebettete Modell vorzeitig beenden.
fn fuer_script_tag(json: &str) -> String {
    json.replace('<', "\\u003c")
}

fn groesse(bytes: usize) -> String {
    if bytes < 1024 {
        format!("{bytes} Bytes")
    } else if bytes < 1024 * 1024 {
        format!("{} KB", bytes / 1024)
    } else {
        format!("{:.1} MB", bytes as f64 / (1024.0 * 1024.0))
    }
}
